api/src/main/java/org/infinispan/api/Infinispan.java

```diff
@@ -14,37 +17,75 @@
  * @since 10.0
  */
 @Experimental("This is not ready yet for general consumption. Major changes are still expected.")
-public interface Infinispan {
-
-   static Infinispan newClient(ClientConfig config) {
-      try {
-         return (Infinispan) Infinispan.class.getClassLoader()
-               .loadClass("org.infinispan.api.client.impl.InfinispanClientImpl")
-               .getConstructor(ClientConfig.class)
-               .newInstance(config);
-      } catch (Exception e) {
+public interface Infinispan extends AutoCloseable {
+   /**
+    * <ul>
+    *   <li><tt>file:///.../infinispan.xml</tt> Embedded Infinispan configured via XML/JSON/Yaml file</li>
+    *   <li><tt>classpath:///.../infinispan.xml</tt> Embedded Infinispan configured via XML/JSON/Yaml classpath resource</li>
+    *   <li><tt>hotrod[s]://[username[:password]@]host:port[,host2:port]?property=value[&property=value]</tt> </li>
+    * </ul>
+    *
+    * @param uri one of the supported Infinispan URIs:
+    * @return an
+    */
+   static Infinispan fromURI(URI uri) {
+      Iterator<Factory> factories = ServiceLoader.load(Factory.class).iterator();
+      while (factories.hasNext()) {
+         Factory factory = factories.next();
+         Infinispan instance = factory.fromURI(uri);
+         if (instance != null) {
+            return instance;
+         }
       }
-      return null;
+      throw new InfinispanConfigurationException("No factory to handle URI " + uri);
+   }
+
+   static Infinispan fromURI(String uri) {
+      return fromURI(URI.create(uri));
    }
 
-   static Infinispan newEmbedded(EmbeddedConfig config) {
-      throw new UnsupportedOperationException("Embedded mode not yet supported");
+   static Infinispan fromConfiguration(Configuration configuration) {
```

Comment:
IMO method overloading is nicer here so that all have the signature `static Infinispan from(...)`. That does lead to the new question though of whether `.new(..)` or `.create(...)` is maybe more appropriate?

Reply:
`new` is a reserved keyword. `create()` is great.

Reply:
> `new` is a reserved keyword

The coffee was still digesting :sweat_smile:

Reply:
I've used `create()` for caches and locks. For counters, since we have two types, I've left `createWeakCounter`